crates/primitives/src/chain_spec.rs

```diff
@@ -107,61 +114,103 @@ impl ChainSpec {
     }
 
     /// Returns the supported hardforks and their fork block numbers
-    pub fn hardforks(&self) -> &BTreeMap<Hardfork, BlockNumber> {
+    pub fn hardforks(&self) -> &BTreeMap<Hardfork, ForkKind> {
         &self.hardforks
     }
 
     /// Get the first block number of the hardfork.
     pub fn fork_block(&self, fork: Hardfork) -> Option<BlockNumber> {
+        self.hardforks
+            .get(&fork)
+            .and_then(|kind| match kind {
+                ForkKind::Block(number) => Some(*number),
+                ForkKind::TTD(ttd) => ttd.block,
+                _ => None,
+            })
+    }
+
+    /// Get the first block number of the hardfork.
+    pub fn fork_kind(&self, fork: Hardfork) -> Option<ForkKind> {
         self.hardforks.get(&fork).copied()
     }
 
     /// Returns `true` if the given fork is active on the given block
-    pub fn fork_active(&self, fork: Hardfork, current_block: BlockNumber) -> bool {
-        self.fork_block(fork).map(|target| target <= current_block).unwrap_or_default()
+    pub fn fork_active(&self, fork: Hardfork, discriminant: ForkDiscriminant) -> bool {
+        match self.hardforks.get(&fork) {
+            Some(kind) => match kind {
+                ForkKind::Block(block_number) => *block_number <= discriminant.block_number,
+                ForkKind::TTD(ttd) => {
+                    ttd.total_difficulty <= discriminant.total_difficulty &&
+                        ttd.block <= Some(discriminant.block_number)
+                }
+                ForkKind::Time(timestamp) => *timestamp <= discriminant.timestamp,
+            },
+            None => false,
+        }
     }
 
     /// Returns `true` if the DAO fork is supported
     pub fn dao_fork_support(&self) -> bool {
         self.dao_fork_support
     }
 
-    /// Get the Paris status
-    pub fn paris_status(&self) -> ParisStatus {
-        match self.paris_ttd {
-            Some(terminal_total_difficulty) => {
-                ParisStatus::Supported { terminal_total_difficulty, block: self.paris_block }
+    /// The merge terminal total difficulty
+    pub fn terminal_total_difficulty(&self) -> Option<U256> {
```

Comment:
This is an OK helper I think, if the assumption is that TTD was only for the merge and never again

Reply:
yeah, this lgtm